Suppress missing reference time warnings

sawjlab opened the pull request:
Coincidence data files may include prescaled singles events. For a singles event for a given spectrometer, the other spectrometer's modules will still be gated and read out. However, the reference times for that other spectrometer will typically not be present. The decoder (DecodeToHitList) will print warning messages for missing reference times. These can be avoided by telling each spectrometer object which event types should include data from that spectrometer. When a spectrometer "X" is not expected to have data (as determined by the event type), the global variable X.present will be false. Each detector class looks at this variable and suppresses the decoder warnings when it is false.

As an example, if "HMS" is the THcHallCSpectrometer object for the high momentum spectrometer and event type 1 is a coincidence event and event type 3 is an HMS singles event, then this warning suppression is setup with:

HMS->AddEvtType(1);
HMS->AddEvtType(3);

(Note: in this example, HMS must be of the type THcHallCSpectrometer*, not THaApparatus*)

If no event types are added for the spectrometer, then all missing reference time warnings are printed.

sawjlab committed:
Flag events as present if evtype is in a list for given spectrometer

sawjlab committed:
Add optional flag to THcHitList::DecodeToHitList to suppress warnings

sawjlab committed:
Suppress missing ref time messages when ref time not expected
  Each detector object will check the global variable X.present each
  event.  If X.present is false, it indicates that the parent
  spectrometer object determineted that this spectrometer is not in
  this event, and therefor reference times are not expected.  If
  X.present is false, then the reference time warning messages will
  be suppressed.

sawjlab commented:
@MarkKJones suggested rephrasing what is happening here a bit.  For those event type that are specified by THcHallCSpectrometer::AddEvtType, warnings about missing reference times will be printed out.

If no event types are specified.  Then warnings will be printed for all event types.

MarkKJones commented:
The trigger apparatus and detector was left out.

sawjlab commented:
I presume that the trigger apparatus has information from both spectrometers and presumably reference times for both spectrometers.  Not sure how to deal with this unless we have separate trigger apparati  for each spectrometer.  We could just suppress missing ref time warnings in the trigger detector for all events.

MarkKJones commented:
I think that we can have two separate TrigDet for HMS and SHMS in the TrigApp. 
Presently for coin script we just have one. It probably makes sense to do this separation anyway.

sawjlab committed:
Suppress missing ref time messages for THcTrigDet when not expected.
  To do this suppression, a trigger detector object needs to know the
  name of spectrometer it is associated with.  By default it will get
  this name from the first character of the it's own name.  This can
  be overridden with the SetSpectName method.
